pickler for GraphModule

aorenste opened the pull request:
Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.




Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141659



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D68921318](https://our.internmc.facebook.com/intern/diff/D68921318)

aorenste committed:
pickler for GraphModule

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ezyang commented:
This is pretty long. What's the best way to approach the review?

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
> This is pretty long. What's the best way to approach the review?

I added some docs which should hopefully make it more approachable.

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1128,6 +1142,194 @@ def log_graph_runnable() -> str:
+        return result
+
+
+class _FxCompileSerialized(FxCompile):
```
What's the definition of a the class FxCompile? Is it related at all to an OutputCode?

jamesjwu replied:
Ah I see. I think this concept is a bit confusing of a design because the nomenclature makes me think we're serializing an object called FxCompile. What it's actually doing is running CompileFx in a subprocess or remote worker. 

Maybe something like _FxCompileExternal would be a better name? Will let you bikeshed though, not a strong opinion

aorenste replied:
`_SerializedFxCompile`?

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1128,6 +1142,194 @@ def log_graph_runnable() -> str:
+        extra_env: Optional[Mapping[str, str]] = None,
+    ) -> _WireProtocolPickledOutput:
+        with contextlib.ExitStack() as stack:
+            if extra_env is not None:
```
maybe unittest_env is a better name for this? Otherwise to me it sounds like you are adding extra environment variables in production

aorenste replied:
There's no reason we couldn't use this to add extra environment vars in production if necessary. It's just extra environment vars set in the child's environment.

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1128,6 +1142,194 @@ def log_graph_runnable() -> str:
+                inputs_to_check,
+                graph_kwargs,
+            ).serialize()
+        except (AttributeError, BypassFxGraphCache):
```
Why is BypassFxGraphCache specifically here? Preferably we have one exception to handle here

jamesjwu commented on torch/_inductor/compile_fx.py:
```diff
@@ -1128,6 +1142,194 @@ def log_graph_runnable() -> str:
+    gm: torch.fx.GraphModule
+    example_inputs: Sequence[InputType]
+    inputs_to_check: Sequence[int]
+    graph_kwargs: _CompileFxKwargs
```
Maybe add a TODO here that adds some arbitrary serializable metadata or logging output by design, so we can always attach metadata to what we're serializing.

aorenste replied:
Lol - in the next PR I actually have the comment there:
```
# TODO: Ugh - what other random state are we missing?
```

jamesjwu commented on torch/fx/_graph_pickler.py:
```diff
@@ -0,0 +1,568 @@
+    """
+
+    def __init__(self, file: io.BytesIO) -> None:
+        super().__init__(file, protocol=pickle.HIGHEST_PROTOCOL)
```
This makes it so that upgrades to pickle/python version break BC, right? Is that okay? Is there a more stable protocol choice?

aorenste replied:
My original thinking was that the pickling was ephemeral so highest would be best, but actually there's no reason someone couldn't insert a cache and make it longer-lasting.

jamesjwu commented on torch/fx/_graph_pickler.py:
```diff
@@ -0,0 +1,568 @@
+        #      support.
+
+        # These are the types that need special handling. See the individual
+        # *PickleData classes for details on pickling that particular type.
```
- What's the recursive behavior of this interface? Let's say I pickle a FakeTensor and it goes to TensorPickleData.reduce_helper. In the meta of that FakeTensor there's a sym int or shape env. Does it call back into this reducer_override? Similar question if something goes wrong and we fallback to default pickling: let's say we find a class Foo that we don't know how to pickle, and we fall back tot ehd efault pickler. Foo then contains another FakeTensor inside of it. Does it call reducer_override again? Or is it permanently default after that? 


- What about real tensors? Sometimes a graph module will have a real tensor embedded in it as a constant. Do we want a pickler for that?

aorenste replied:
Yes - the reducer_override is recursively applied on the output.
Real tensors - maybe? The problem with real tensors is that it's not obvious we can unpickle them (and we'd need to pickle all the data which could be huge)

jamesjwu replied:
Hm there's got to be a unit test that exhibits adding a real tensor as a constant in the graph module... is the behavior today that we would throw NotImplemented and use the default pickler?

aorenste replied:
Correct - for any type that isn't explicitly handled here we *return* NotImplemented (not thrown!) which uses the default behavior.

jamesjwu reviewed (commented):
What's your thoughts on testing individual picklers? We have a single unit test right now that pickles a really basic graph, but in order for this pickler to actually be ready I think we kind of need to at least run it on every unit test, and preferably benchmarks as well. That way, we can get a sense of various serializations that need to be implemented.

I would also love to use GraphPickler to serialize backward graph modules in AOTAutogradCache, to support things like CompiledAutograd (cc @bdhirsh). Hopefully that should be easy enough to support (even if I have to add some more custom picklers). It would also make lazy backward much easier to support, since we could pickle the input to FXGraphCache instead of waiting for backward to complete.

aorenste committed:
None on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jamesjwu commented on torch/_dynamo/variables/misc.py:
```diff
@@ -1220,6 +1220,12 @@ def get_np_to_tnp_map():
 
 
+@functools.lru_cache(maxsize=1)
+def get_tnp_to_np_map():
```
nit: As someone unfamiliar with this stuff this function is extraordinarily mysterious lol

aorenste replied:
Added a docstring.

jamesjwu commented on test/inductor/test_torchinductor.py:
```diff
@@ -13026,9 +13029,11 @@ def fn3(x):
-                    "triton_poi_fused_layer_norm_relu"
-                    if torch._dynamo.config.inline_inbuilt_nn_modules
-                    else "triton_poi_fused_LayerNorm_ReLU",
+                    (
```
Is there any reason for the changes to this file? They seem unintentional and mess up the blame

aorenste replied:
I think it was just the linter jumping in and auto-fixing it. I'll remove them.

jamesjwu commented on torch/fx/_graph_pickler.py:
```diff
@@ -0,0 +1,582 @@
+            context.force_unspec_int_unbacked_size_like
+        )
+        # Not saved (because it's difficult and maybe not needed?):
+        #   self.fw_metadata = context.fw_metadata
```
It's hard for me to reason about the intentions here. Is the goal that when unpickling a graph module that we also reinstate the exact tracing context that it came from?

A lot of these seem specific to the tracing step that uses the graph module, instead of being "owned" by the graph module itself. Do you have a sense of how the GraphModule uses the fields in this tracingcontext (At least, the ones that you do end up pickling and saving)? 

I'm just a bit fuzzy on what exactly this context is doing

aorenste replied:
The intentions were that I just wanted to save what was needed and no more. (There's a lot of "junk" hanging off some of our structures). I listed what I didn't save so in case we're having a problem and grep for it we'll see that it's explicitly not saved (and maybe that should be fixed).

jamesjwu reviewed (approved):
Seems pretty reasonable to me, just some minor nits and questions

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge

aorenste commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Sorry but this is breaking internally, please take a look at D68694181 for more details."

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68785860).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68785860).

aorenste committed:
Update on "pickler for GraphModule"

Pickling GraphModule needs some special handling for wrapping things that normally can't be pickled - but async compile needs to pass them across a wire so we need to be able to serialize it - add some helpers to enable that.







cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D68785860](https://our.internmc.facebook.com/intern/diff/D68785860)

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68921318).

aorenste commented:
@pytorchbot merge
